Implementing JSON Schema validation

lahmatiy opened the pull request:
Initial implementation of JSON Schema validation (fixes #19).

At the moment all `css/*.json` files are correct except `properties.json`. Last one has several issues that were addressed by PRs ~#62~, ~#64~, ~#65~, ~#68~, ~#70~, ~#73~, ~#74~, ~#75~, ~#76~, ~#82~ and ~#85~.

Also I added `npm test` script. I'm going to integrate it to travis and merge its current script with it later.

a2sheppy commented:
This is really exciting!

lahmatiy commented:
PR is ready for merge.
There is a single issue with `computed` value of `box-suppress` property (warnings with `animationType` are addressed in #70). Currently its value is `seeProse`. I think it's wrong and should be replaced for proper value. But last [changes in spec](https://drafts.csswg.org/css-display/#changes) is deferred this property, so no any spec about property and any implementation. Should we remove it or fix a dictionary/schema?
//cc @Elchi3

Elchi3 commented:
If there is no spec and no implementation for `box-suppress`, lets remove it entirely. I don't think we want theoretical things in here that no one can use.

I haven't reviewed the schema yet. I will try to come to that soon. If you could add the test script also to the .travis.yml, that would be nice!

lahmatiy commented:
> If there is no spec and no implementation for box-suppress, lets remove it entirely

See #82 

> If you could add the test script also to the .travis.yml, that would be nice!

Done! Now Travis validates JSON schema along to JSON linting (see https://travis-ci.org/mdn/data/jobs/233813298)

lahmatiy commented:
One missed issue is percentage value for `clip-path`. Issue on spec: https://github.com/w3c/fxtf-drafts/issues/155 (duplicate https://github.com/w3c/fxtf-drafts/issues/168)

Elchi3 commented:
@lahmatiy really nice work! I don't understand the added keyword `property-reference` and `$data` 
What does `"$data": "3"` or `"$data": "/"` mean? I've read some docs about `$data`, but it is still unclear to me.

lahmatiy commented:
@Elchi3 `ajv` allows extend schema syntax. So `property-reference` is my [extension](https://github.com/lahmatiy/data/blob/797983af57bdcc76c970ef8f934ef2ec6cb5568c/test/validate-schema.js#L8-L19) to validate value is a existing key in some object (dictionary). `$data` is [proposed extension](https://github.com/json-schema/json-schema/wiki/%24data-(v5-proposal)) for JSON Schema to make a reference to some part of JSON. It uses using [JSON Pointers](https://tools.ietf.org/html/rfc6901) or [Relative JSON Pointers](http://tools.ietf.org/html/draft-luff-relative-json-pointer-00) as a value.
All this stuff used in two cases:
- `css/at-rules.schema.json` – check value from an array is existing key in `descriptors` dictionary. `"$data": "3"` is relative JSON pointer, means that a three level up value should be used as dictionary. For instance:
```
{
  ...
  "@viewport": {
    ...
    "descriptors": {
      "height": {
        "syntax": "<viewport-length>{1,2}",
        "media": "visual, continuous",
        "initial": [
          "min-height",
          "max-height"
        ]
```
Here `min-height` and `max-height` are checking. Jump 3 level up: `min-height` (0) -> `"initial": [..]` (1) -> `"height": { ... }` (2) -> `"descriptors": { ... }` (3). So `min-height`/`max-height` should be defined in `descriptors`.
- `css/properties.schema.json` – check value is existing property (defined on top level of dictionary). `"$data": "/"` is JSON pointer that points to root of JSON file.

I thought it would be nice to have that kind of validation. We can remove it if there is any reason for that.

lahmatiy commented:
Current state:
- Added a comment to JSON schema for less confusing about extension
- Updated branch from upstream
- Made pull requests #85 and #86 to fix remaining issues

Elchi3 commented on css/at-rules.schema.json:
```diff
@@ -12,6 +12,7 @@
           "items": {
             "type": "string",
             "property-reference": {
+              "comment": "That's a non-standart extension to JSON Schema validator. It tests value is an existing key in `descriptors`. See test/validate-schema.js for extension implementation details",
```
Better:
"This is an extension to the JSON schema validator. It jumps 3 levels up in the hierarchy and tests if a value is an existing key in `descriptors`. See test/validate-schema.js for implementation details."

Elchi3 commented on css/properties.schema.json:
```diff
@@ -7,6 +7,7 @@
       "items": {
         "type": "string",
         "property-reference": {
+          "comment": "That's a non-standart extension to JSON Schema validator. It tests value is an existing key in root object (i.e. defined property). See test/validate-schema.js for extension implementation details",
```
Better:
"property-reference is an extension to the JSON schema validator. Here it jumps to the root level of the hierarchy and tests if a value is an existing key there (i.e a defined property). See test/validate-schema.js for implementation details."

lahmatiy commented:
@Elchi3 Thank you for quick review!
Comments are fixed, branch updated from upstream and related changes are made.
All checks have passed 🎉
